crates/services/p2p/src/discovery.rs

```diff
@@ -184,51 +154,20 @@ impl NetworkBehaviour for DiscoveryBehaviour {
             }
         }
 
-        // poll Kademlia behaviour
-        while let Poll::Ready(kad_action) = self.kademlia.poll(cx, params) {
+        // poll sub-behaviors
+        if let Poll::Ready(kad_action) = self.kademlia.poll(cx, params) {
             match kad_action {
-                NetworkBehaviourAction::GenerateEvent(
-                    KademliaEvent::UnroutablePeer { peer },
-                ) => {
-                    return Poll::Ready(NetworkBehaviourAction::GenerateEvent(
-                        DiscoveryEvent::UnroutablePeer(peer),
-                    ))
-                }
-
-                NetworkBehaviourAction::Dial { handler, opts } => {
-                    return Poll::Ready(NetworkBehaviourAction::Dial { handler, opts })
-                }
-                NetworkBehaviourAction::CloseConnection {
-                    peer_id,
-                    connection,
-                } => {
-                    return Poll::Ready(NetworkBehaviourAction::CloseConnection {
-                        peer_id,
-                        connection,
-                    })
-                }
-                NetworkBehaviourAction::NotifyHandler {
-                    peer_id,
-                    handler,
-                    event,
-                } => {
-                    return Poll::Ready(NetworkBehaviourAction::NotifyHandler {
-                        peer_id,
-                        handler,
-                        event,
-                    })
-                }
-                NetworkBehaviourAction::ReportObservedAddr { address, score } => {
-                    return Poll::Ready(NetworkBehaviourAction::ReportObservedAddr {
-                        address,
-                        score,
-                    })
-                }
-                _ => {}
+                NetworkBehaviourAction::GenerateEvent(event) => match event {
+                    KademliaEvent::UnroutablePeer { peer } => {
+                        Poll::Ready(NetworkBehaviourAction::GenerateEvent(
+                            KademliaEvent::UnroutablePeer { peer },
+                        ))
+                    }
+                    _ => Poll::Pending,
+                },
+                _ => Poll::Ready(kad_action),
```

Comment:
I had that before, but technically that's not what the code was doing before because it was only matching on the one event variant.

I can switch it if you think it's safe (I think it is too).